src/optimization/ControlFlow.cpp

```diff
@@ -971,3 +974,87 @@ void optimizations::addStartStopSegment(const Module& module, Method& method, co
 
 	generateStopSegment(method);
 }
+
+LoopInclusionTreeNode* findRoot(LoopInclusionTreeNode *node)
+{
+	for (auto &parent : node->getNeighbors())
+	{
+		if (!parent.second.includes) {
+			// The root node must be only one
+			return findRoot(parent.first);
+		}
+	}
+	// this is root
+	return node;
+}
+using LoopInclusionTree = Graph<ControlFlowLoop*, LoopInclusionTreeNode>;
+
+void optimizations::removeConstantLoadInLoops(const Module& module, Method& method, const Configuration& config)
+{
+	// 1. find loops
+	auto cfg = ControlFlowGraph::createCFG(method);
+	auto loops = cfg.findLoops();
+
+	// 2. generate inclusion relation of loops as trees
+	LoopInclusionTree inclusionTree;
+	for (auto &loop1 : loops)
+	{
+		for (auto &loop2 : loops)
+		{
+			if (loop1 == loop2) continue;
+
+			if (loop1.includes(loop2))
+			{
+				auto &node1 = inclusionTree.getOrCreateNode(&loop1);
+				auto &node2 = inclusionTree.getOrCreateNode(&loop2);
+				node1.addNeighbor(&node2, LoopInclusion(true));
+				node2.addNeighbor(&node1, LoopInclusion(false));
+			}
+		}
+	}
+
+	logging::debug() << "loops" << logging::endl;
+	for (auto &loop : loops) {
+		logging::debug() << "  " << &loop << logging::endl;
+	}
+
+	logging::debug() << "inclusionTree" << logging::endl;
+	for (auto &loop : inclusionTree) {
+		logging::debug() << "  " << loop.first << logging::endl;
+		for (auto &node : loop.second.getNeighbors()) {
+			logging::debug() << "    " << node.first->key << ": " << node.second.includes << logging::endl;
+		}
+	}
+
+	// 3. move constant load operations from root of trees
+	FastSet<ControlFlowLoop*> processed;
+	for (auto &loop : loops)
+	{
+		logging::debug() << "loop: " << &loop << logging::endl;
+		auto &node = inclusionTree.getOrCreateNode(&loop);
+		auto root = findRoot(&node);
+
+		if (processed.find(root->key) != processed.end()) continue;
+		processed.insert(root->key);
+
+		for (auto &cfgNode : *root->key)
+		{
+			auto block = cfgNode->key;
+			for (auto it = block->begin(); it != block->end(); it = it.nextInBlock())
+			{
+				// TODO: Constants like `mul24 r1, 4, elem_num` should be also moved.
+				auto loadInst = it.get<LoadImmediate>();
+				if (loadInst != nullptr)
+				{
+					// LoadImmediate must have output value
+					auto out = loadInst->getOutput().value();
+					if (out.valueType == ValueType::LOCAL) {
+						auto rootInst = method.begin()->end().previousInBlock();
```

Comment:
Here you hoist every constant load up to the end of the first basic block.
It should be enough to hoist them to the end of the block directly preceding the loop (e.g. via `loop.findPredecessor()`). This way, we put less pressure on the register-allocator later on.